test/jobs/TestJobEngine.cpp

```diff
@@ -0,0 +1,161 @@
+{
+    vector<PlainJobDocument::JobAction> steps;
+    vector<std::string> args;
+    steps.push_back(
+        createJobAction("testAction", "runHandler", "successHandler", args, "/tmp/device-client-tests/", false));
+    PlainJobDocument jobDocument = createTestJobDocument(steps, true);
+    JobEngine jobEngine;
+
+    int executionStatus = jobEngine.exec_steps(jobDocument, testHandlerDirectoryPath);
+    ASSERT_EQ(executionStatus, 0);
+    ASSERT_STREQ(jobEngine.getStdOut().c_str(), std::string(testStdout + "\n").c_str());
+}
+
+TEST_F(TestJobEngine, ExecuteSucceedThenFail)
+{
+    vector<PlainJobDocument::JobAction> steps;
+    vector<std::string> args;
+    steps.push_back(
+        createJobAction("testAction", "runHandler", "successHandler", args, "/tmp/device-client-tests/", false));
+    PlainJobDocument::JobAction finalStep =
+        createJobAction("testAction", "runHandler", "errorHandler", args, "/tmp/device-client-tests/", false);
+    PlainJobDocument jobDocument = createTestJobDocument(steps, finalStep, true);
+    JobEngine jobEngine;
+
+    int executionStatus = jobEngine.exec_steps(jobDocument, testHandlerDirectoryPath);
+    ASSERT_NE(executionStatus, 0);
+    ASSERT_STREQ(jobEngine.getStdOut().c_str(), std::string(testStdout + "\n").c_str());
+    ASSERT_STREQ(jobEngine.getStdErr().c_str(), std::string(testStderr + "\n").c_str());
+}
+
+TEST_F(TestJobEngine, ExecuteFinalStepOnly)
+{
+    vector<PlainJobDocument::JobAction> steps;
+    vector<std::string> args;
+    PlainJobDocument::JobAction finalStep =
+        createJobAction("testAction", "runHandler", "successHandler", args, "/tmp/device-client-tests/", false);
+    PlainJobDocument jobDocument = createTestJobDocument(steps, finalStep, true);
+    JobEngine jobEngine;
+
+    int executionStatus = jobEngine.exec_steps(jobDocument, testHandlerDirectoryPath);
+    ASSERT_EQ(executionStatus, 0);
+    ASSERT_STREQ(jobEngine.getStdOut().c_str(), std::string(testStdout + "\n").c_str());
+}
+
+TEST_F(TestJobEngine, ExecuteStepsError)
+{
+    vector<PlainJobDocument::JobAction> steps;
+    vector<std::string> args;
+    steps.push_back(
+        createJobAction("testAction", "runHandler", "errorHandler", args, "/tmp/device-client-tests/", false));
+    PlainJobDocument jobDocument = createTestJobDocument(steps, true);
+    JobEngine jobEngine;
+
+    int executionStatus = jobEngine.exec_steps(jobDocument, testHandlerDirectoryPath);
+    ASSERT_NE(executionStatus, 0);
+    ASSERT_STREQ(jobEngine.getStdErr().c_str(), std::string(testStderr + "\n").c_str());
+}
+
+TEST_F(TestJobEngine, ExecuteNoSteps)
+{
+    vector<PlainJobDocument::JobAction> steps;
+    vector<std::string> args;
+    PlainJobDocument jobDocument = createTestJobDocument(steps, true);
+    JobEngine jobEngine;
+
+    int executionStatus = jobEngine.exec_steps(jobDocument, testHandlerDirectoryPath);
+    ASSERT_EQ(executionStatus, 0);
+    ASSERT_TRUE(jobEngine.getStdOut().length() == 0);
```

Comment:
I'll change this to ASSERT_EQ(..., 0) -- I was going to have it be .empty() then didn't fix here.